taichi/codegen/codegen_opengl.cpp

```diff
@@ -95,18 +86,51 @@ struct CompiledProgram {
     auto gtmp_arr = std::vector<char>(gtmp_size);
     void *gtmp_base = gtmp_arr.data();  // std::calloc(gtmp_size, 1);
     iov.push_back(IOV{gtmp_base, gtmp_size});
-    if (has_ext_arr) {
+    std::optional<std::vector<char>> base_arr;
+    std::optional<std::vector<void *>> saved_ctx_ptrs;
+    // TODO: these dirty codes are introduced by #694
+    if (ext_arr_map.size()) {
       iov.push_back(
           IOV{ctx.extra_args, arg_count * taichi_max_num_args * sizeof(int)});
-      void *extptr = (void *)ctx.args[ext_arr_idx];
-      ctx.args[ext_arr_idx] = 0;
-      iov.push_back(IOV{extptr, ext_arr_size});
+      if (ext_arr_map.size() == 1) { // zero-copy for only one ext_arr
+        auto it = ext_arr_map.begin();
+        auto extptr = (void *)ctx.args[it->first];
+        ctx.args[it->first] = 0;
+        iov.push_back(IOV{extptr, it->second});
+      } else {
+        size_t accum_size = 0;
+        std::vector<void *> ptrarr;
+        for (auto it = ext_arr_map.begin(); it != ext_arr_map.end(); it++) {
```

Comment:
Range-based for loops are your friend: https://en.cppreference.com/w/cpp/language/range-for

 Are h you can probably write `for (const auto &it: ext_arr_map)`

Reply:
Cool!!